asterius/rts/rts.gc.mjs

```diff
@@ -120,6 +120,117 @@ export class GC {
+    // closures, until it finds a result. It proceeds in two phases:
+    // - Search phase: indirectee/selectee pointers are followed,
+    //     and the closures encountered are pushed on a stack
+    // - Back propagation phase: the closures on the stack are popped
+    //     and replaced with IND closures where possible.
+    // Note: the two phases can be interleaved. 
+    let result, peek;
+    while (stack.length) {
+      if (!result) {
+        // Search phase
+        [c, untagged_c, info, type] = peek = stack[stack.length-1];
+        if (!untagged_c) {
+          // If no information about c is present, compute it
+          const untagged_c = peek[1] = Memory.unDynTag(c),
+            info = peek[2] = Number(this.memory.i64Load(untagged_c));
+          if (info % 2) {
+            // The closures has already been overwritten with a 
+            // forwarding pointer: switch phase.
+            result = stack.pop(); // = peek
+            continue;
+          }
+          type = peek[3] = this.memory.i32Load(info + rtsConstants.offset_StgInfoTable_type);
+        }
+        switch (type) {
+          case ClosureTypes.IND: {
+            // Whitehole
+            this.memory.i64Store(untagged_c, this.symbolTable.stg_WHITEHOLE_info);
+            // Follow the indirectee
+            stack.push([this.memory.i64Load(untagged_c + rtsConstants.offset_StgInd_indirectee),,,]);
+            continue;
+          }
+          case ClosureTypes.THUNK_SELECTOR: {
+            // Whitehole
+            this.memory.i64Store(untagged_c, this.symbolTable.stg_WHITEHOLE_info);
+            // Follow the selectee
+            stack.push([this.memory.i64Load(untagged_c + rtsConstants.offset_StgSelector_selectee),,,]);
+            continue;
+          }
+          default: {
+            // Switch phase
+            result = stack.pop();
+            continue;
+          }
+        }
+      } else {
+        // Back propagation phase
+        [c, untagged_c, info, type] = peek = stack.pop();
+        let [res_c, _, res_info, res_type] = result;
+        switch (type) {
+          case ClosureTypes.IND: {
+            if (res_info % 2) {
+              this.memory.i64Store(untagged_c, res_info);
+              info = res_info; type = res_type;
+            } else {
+              this.memory.i64Store(untagged_c, this.symbolTable.stg_IND_info); // Undo whiteholing
+              this.memory.i64Store(untagged_c + rtsConstants.offset_StgInd_indirectee, res_c);
+            }
+            continue;
+          }
+          case ClosureTypes.THUNK_SELECTOR: {
+            if (res_type == ClosureTypes.CONSTR) {
+              const offset = this.memory.i32Load(
+                res_info + rtsConstants.offset_StgInfoTable_layout
+              );
+              peek[2] = this.symbolTable.stg_IND_info;
+              stack.push(peek); 
+              // Start evaluating the selected field
+              stack.push([res_c + offset,,,]);
```

Comment:
```javascript
              stack.push([res_c + offset,,,]);
```

What's this `,,,` syntax?

Reply:
`[res_c + offset,,,]` means an array with four entries, the last three are `undefined`. I made explicit the `undefined` entries just because only arrays that are quadruples are pushed to the `stack`.

Reply:
Ah, thanks for the explanation!